aws_lambda_powertools/utilities/data_classes/s3_batch_operation_event.py

```diff
@@ -0,0 +1,220 @@
+    - https://docs.aws.amazon.com/AmazonS3/latest/API/API_control_LambdaInvokeOperation.html#AmazonS3-Type-control_LambdaInvokeOperation-InvocationSchemaVersion
+
+    Parameters
+    ----------
+    invocation_schema_version : str
+        Specifies the schema version for the payload that Batch Operations sends when invoking
+        an AWS Lambda function., either '1.0' or '2.0'. This must be copied from the event.
+
+    invocation_id : str
+        The identifier of the invocation request. This must be copied from the event.
+
+    treat_missing_keys_as : Literal["Succeeded", "TemporaryFailure", "PermanentFailure"]
+        undocumented parameter, defaults to "PermanentFailure"
+
+    results : List[S3BatchOperationResult]
+        results of each S3 Batch Operations task,
+        optional parameter at start. can be added later using `add_result` function.
+
+    Examples
+    --------
+
+    **S3 Batch Operations**
+
+    ```python
+    from aws_lambda_powertools.utilities.typing import LambdaContext
+    from aws_lambda_powertools.utilities.data_classes import (
+        S3BatchOperationEvent,
+        S3BatchOperationResponse,
+        S3BatchOperationResult
+    )
+
+    def lambda_handler(event: dict, context: LambdaContext):
+        s3_event = S3BatchOperationEvent(event)
+        response = S3BatchOperationResponse(s3_event.invocation_schema_version, s3_event.invocation_id)
+        result = None
+
+        task = s3_event.task
+        try:
+            do_work(task.s3_bucket, task.s3_key)
+            result = S3BatchOperationResult.as_succeeded(task)
+        except TimeoutError as e:
+            result = S3BatchOperationResult.as_temporary_failure(task, str(e))
+        except Exception as e:
+            result = S3BatchOperationResult.as_permanent_failure(task, str(e))
+        finally:
+            response.add_result(result)
+
+        return response.asdict()
+    ```
+    """
+
+    invocation_schema_version: str
+    invocation_id: str
+    treat_missing_keys_as: Literal["Succeeded", "TemporaryFailure", "PermanentFailure"] = "PermanentFailure"
+    results: List[S3BatchOperationResult] = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.treat_missing_keys_as not in VALID_RESULT_CODE_TYPES:
+            raise ValueError(f"Invalid treat_missing_keys_as: {self.treat_missing_keys_as}")
+
+    def add_result(self, result: S3BatchOperationResult):
+        self.results.append(result)
+
+    def asdict(self) -> Dict:
+        if not self.results:
+            raise ValueError("Response must have one result")
+        if len(self.results) > 1:
+            raise ValueError("Response cannot have more than one result")
+
+        return {
+            "invocationSchemaVersion": self.invocation_schema_version,
+            "treatMissingKeysAs": self.treat_missing_keys_as,
+            "invocationId": self.invocation_id,
+            "results": [result.asdict() for result in self.results],
+        }
```

Comment:
```suggestion
    def asdict(self) -> Dict:

        result_count = len(self.results)

        if result_count != 1:
            raise ValueError(f"Response must have exactly one result, but got {result_count}")

        return {
            "invocationSchemaVersion": self.invocation_schema_version,
            "treatMissingKeysAs": self.treat_missing_keys_as,
            "invocationId": self.invocation_id,
            "results": [result.asdict() for result in self.results],
        }
```

We can make it simpler and a little more performant.